### PR TITLE
Fix LoadConfiguration for not found config file

### LINKED ISSUE #2878: TryLoadLoggingConfiguration throws exception instead of just logging failure.
**NLog version**: 4.5.9

**Platform**: .Net 4.6.1

**Current NLog config** (xml or C#, if relevant)
None

- What is the current result?
 NLog.LogFactory.GetLogger throws an exception if it can't find an nlog.config file.

- What is the expected result?
No exception, just no configuration loaded.  "Failed loading config from ..." logged to InternalLogger.

- Did you checked the [Internal log](https://github.com/NLog/NLog/wiki/Internal-Logging)?
Yes

- Please post full exception details (message, stacktrace, inner exceptions)
Note: Private details changed to ... .
``` C#
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. 
---> System.TypeInitializationException: The type initializer for '...' threw an exception. 
---> System.TypeInitializationException: The type initializer for '...' threw an exception. 
---> System.TypeInitializationException: The type initializer for '...' threw an exception. 
---> System.IO.FileNotFoundException: Could not find file '...NLog.config'.
    at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, String msgPath, Boolean bFromProxy)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize)
   at System.Xml.XmlDownloadManager.GetStream(Uri uri, ICredentials credentials, IWebProxy proxy, RequestCachePolicy cachePolicy)
   at System.Xml.XmlUrlResolver.GetEntity(Uri absoluteUri, String role, Type ofObjectToReturn)
   at System.Xml.XmlTextReaderImpl.FinishInitUriString()
   at System.Xml.XmlTextReaderImpl..ctor(String uriStr, XmlReaderSettings settings, XmlParserContext context, XmlResolver uriResolver)
   at System.Xml.XmlReaderSettings.CreateReader(String inputUri, XmlParserContext inputContext)
   at System.Xml.XmlReader.Create(String inputUri, XmlReaderSettings settings, XmlParserContext inputContext)
   at System.Xml.XmlReader.Create(String inputUri)
   at NLog.Config.XmlLoggingConfiguration.CreateFileReader(String fileName)
   at NLog.Config.XmlLoggingConfiguration..ctor(String fileName, Boolean ignoreErrors, LogFactory logFactory)
   at NLog.Config.XmlLoggingConfiguration..ctor(String fileName, LogFactory logFactory)
   at NLog.LogFactory.TryLoadLoggingConfiguration(String configFile)
   at NLog.LogFactory.TryLoadFromFilePaths()
   at NLog.LogFactory.get_Configuration()
   at NLog.LogFactory.GetLogger(LoggerCacheKey cacheKey)
   at NLog.LogFactory.GetLogger(String name)
   at NLog.LogManager.GetLogger(String name)
   at ...  

```

- Are there any workarounds? yes
Add nlog.config that is empty.

- Is there a version in which it did work?
Maybe 4.5.7... removed a bad nlog.config file from a dependency that was being copied into the folder.

- Can you help us by writing an unit test?
``` C#
[Test]
        public void LogManager_TryLoadLoggingConfiguration_WhenFileDoesNotExist_ThenDoesNotThrowException()
        {
            //Assign
            string filename = $"{Guid.NewGuid()}.config";
            File.Delete(filename);

            //Act
            NLog.LogFactory.TryLoadLoggingConfiguration(filename);            
        }
```

### PR DESCRIPTION
**Pull request by snakefoot**
TryLoadFromFilePaths - Improve exception handling

Resolves #2878

### CODE REVIEW

**Comment by 304NotModified** [comment]
so `factory.LoadConfiguration(filename);` still fails even when throwError = false? (see #2880)

**Comment by snakefoot** [comment]
The difference is that `LoadConfiguration` says load this specific file at this expected path.

While `TryLoadFromFilePaths` just jumps around and loads from all kind of random locations (Without the user can do anything).

**Comment by 304NotModified** [comment]
true, unfortunately the `TryLoadLoggingConfiguration` isn't really a try IMO, as all the check logic is outside of it. But we could refactor that later.

**Comment by snakefoot** [comment]
@304NotModified Now changed TryLoadLoggingConfiguration to return true/false

File: src/NLog/LogFactory.cs [hunk 1]
```diff
@@ -330,7 +323,7 @@ private void TryLoadFromAndroidAssets()
                 {
                     if (stream != null)
                     {
-                        _config = TryLoadLoggingConfiguration(XmlLoggingConfiguration.AssetsPrefix + nlogConfigFilename);
+                        _config = LoadXmlLoggingConfiguration(XmlLoggingConfiguration.AssetsPrefix + nlogConfigFilename);
```
[the remark this answers is not in the document]
Comment (304NotModified) [reply src/NLog/LogFactory.cs]: that's also testcase worthy ;)

Not sure if we could test internal tests in NLog (then we could add then tests to the  TryMethod)

**Comment by snakefoot** [comment]
@304NotModified Created unit-test for you

File: tests/NLog.UnitTests/LogFactoryTests.cs [hunk 2]
```diff
@@ -96,6 +94,32 @@ public void InvalidXMLConfiguration_ThrowErrorWhen_ThrowExceptionFlagIsSet()
...
+            {
+                Directory.CreateDirectory(tempDirectory);
+                File.WriteAllText(configFile, "<nlog />");
+                LogFactory logFactory = new LogFactory();
+                logFactory.SetCandidateConfigFilePaths(new[] { configFile });
+                using (var fileLock = new FileStream(configFile, FileMode.Open, FileAccess.ReadWrite, FileShare.None))
+                {
+                    var loggingConfig = logFactory.Configuration;
+                    Assert.Null(loggingConfig);
+                    Assert.ThrowsAny<Exception>(() => logFactory.LoadConfiguration(configFile));
```
Comment (304NotModified) [inline tests/NLog.UnitTests/LogFactoryTests.cs]: i'm a bit confused. The method title says "DoesNotThrow", but the assert says the contradiction.

**Comment by snakefoot** [comment]
There are two asserts. First it tries to load config by random scan. It will fail and returns null without throwing exception. Next call performs explicit load that should fail by throwing exception


Sent from my Samsung device


-------- Original message --------

File: tests/NLog.UnitTests/LogFactoryTests.cs (hunk 2 above)
Comment (304NotModified) [reply tests/NLog.UnitTests/LogFactoryTests.cs]: Maybe we should split this into 2 unit tests, or rename it. Currently, I don't know what to do with this. I kinda brain overload.

File: src/NLog/LogFactory.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (304NotModified) [reply src/NLog/LogFactory.cs]: PS I try to use these guidelines for unit tests nowadays:

- naming rule 1: https://dzone.com/articles/7-popular-unit-test-naming
-  Arrange, Act, Assert: http://defragdev.com/blog/?p=783

**Comment by snakefoot** [comment]
Well I like to verify that the test-stage actually work while also verifying the code works.

That is why there are two asserts. One for the test of loading random configs and one for the actual stage. Guess I should have a third assert after removing the file lock that both assets now returns not-null



Sent from my Samsung device


-------- Original message --------

File: tests/NLog.UnitTests/LogFactoryTests.cs (hunk 2 above)
Comment (304NotModified) [reply tests/NLog.UnitTests/LogFactoryTests.cs]: I think these are 2 tests.  --> https://github.com/snakefoot/NLog/pull/4

**Comment by 304NotModified** [comment]
see snakefoot#4

**Comment by 304NotModified** [comment]
mmm cherry pick wasn't an option?

**Comment by snakefoot** [comment]
@304NotModified Like I said Github wasn't responding to 50 pct of https-requests

**Comment by 304NotModified** [comment]
the horror

:P

nice:




any way, thanks!

**Comment by 304NotModified** [comment]
note: backported to "dev"

**Pull request merged**